File: prayer_app_flutter/lib/src/screens/quran_bookmarks_screen.dart
import 'package:flutter/material.dart';
import 'package:material_design_icons_flutter/material_design_icons_flutter.dart';
import '../models/quran_models.dart';
import '../services/quran_api_service.dart';
import '../services/quran_storage_service.dart';
import '../theme/app_theme.dart';
import '../providers/theme_provider.dart';
import 'quran_reader_screen.dart';

class QuranBookmarksScreen extends StatefulWidget {
  const QuranBookmarksScreen({super.key});

  @override
  State<QuranBookmarksScreen> createState() => _QuranBookmarksScreenState();
}

class _QuranBookmarksScreenState extends State<QuranBookmarksScreen> {
  final QuranStorageService _storage = QuranStorageService();
  final QuranApiService _api = QuranApiService();

  List<QuranPointer> _bookmarks = [];
  Map<int, SurahMeta> _surahMap = {};
  Map<String, String> _previewByPointer = {};
  bool _loading = true;

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    final bookmarks = await _storage.loadBookmarks();

    List<SurahMeta> surahs = await _api.loadCachedSurahList() ?? [];
    if (surahs.isEmpty) {
      try {
        surahs = await _api.fetchSurahList();
      } catch (_) {}
    }

    if (!mounted) return;
    setState(() {
      _bookmarks = bookmarks;
      _surahMap = {for (final s in surahs) s.number: s};
      _loading = false;
    });

    _loadPreviews(bookmarks);
  }

  Future<void> _loadPreviews(List<QuranPointer> bookmarks) async {
    final next = <String, String>{};
    final cacheBySurah = <int, List<Ayah>>{};

    for (final b in bookmarks) {
      if (!cacheBySurah.containsKey(b.surahNumber)) {
        List<Ayah>? ayahs = await _api.loadCachedArabic(b.surahNumber);
        if (ayahs == null || ayahs.isEmpty) {
          try {
            ayahs = await _api.fetchSurahArabic(b.surahNumber);
          } catch (_) {
            ayahs = [];
          }
        }
        cacheBySurah[b.surahNumber] = ayahs;
      }

      final ayahs = cacheBySurah[b.surahNumber] ?? const <Ayah>[];
      final idx = b.ayahNumber - 1;
      if (idx >= 0 && idx < ayahs.length) {
        final clean = ayahs[idx].textAr.replaceAll('\n', ' ').trim();
        next[b.dedupeKey()] = clean.length > 30 ? '${clean.substring(0, 30)}…' : clean;
      } else {
        next[b.dedupeKey()] = '';
      }
    }

    if (!mounted) return;
    setState(() => _previewByPointer = next);
  }

  Future<void> _open(QuranPointer pointer) async {
    final surah = _surahMap[pointer.surahNumber];
    if (surah == null) return;
    await Navigator.of(context).push(
      MaterialPageRoute(
        builder: (_) => QuranReaderScreen(
          surah: surah,
          initialAyahNumber: pointer.ayahNumber,
        ),
      ),
    );
    _load();
  }

  Future<void> _deleteBookmark(QuranPointer pointer) async {
    await _storage.removeBookmark(pointer);
    _load();
  }

  @override
  Widget build(BuildContext context) {
    final tc = ThemeScope.of(context).current;

    return Scaffold(
      body: Container(
        decoration: BoxDecoration(gradient: appBackgroundGradient(tc)),
        child: SafeArea(
          bottom: false,
          child: Column(
            children: [
              Padding(
                padding: const EdgeInsets.fromLTRB(8, 8, 16, 8),
                child: Row(
                  children: [
                    IconButton(
                      onPressed: () => Navigator.of(context).pop(),
                      icon: Icon(MdiIcons.arrowLeft, color: tc.textPrimary),
                    ),
                    Expanded(
                      child: Text('Bookmarks', style: AppTypography.titleMedium(tc)),
                    ),
                  ],
                ),
              ),
              Expanded(
                child: _loading
                    ? Center(child: CircularProgressIndicator(color: tc.accent))
                    : _bookmarks.isEmpty
                        ? Center(
                            child: Text(
                              'No bookmarks yet',
                              style: AppTypography.caption(tc),
                            ),
                          )
                        : ListView.builder(
                            itemCount: _bookmarks.length,
                            padding: const EdgeInsets.fromLTRB(20, 10, 20, 20),
                            itemBuilder: (context, index) {
                              final b = _bookmarks[index];
                              final surah = _surahMap[b.surahNumber];
                              final preview = _previewByPointer[b.dedupeKey()] ?? '';
                              return GestureDetector(
                                onTap: () => _open(b),
                                onLongPress: () async {
                                  final confirmed = await showModalBottomSheet<bool>(
                                    context: context,
                                    backgroundColor: tc.modalBg,
                                    builder: (_) => SafeArea(
                                      child: Padding(
                                        padding: const EdgeInsets.all(16),
                                        child: Column(
                                          mainAxisSize: MainAxisSize.min,
                                          children: [
                                            ListTile(
                                              leading: Icon(MdiIcons.deleteOutline, color: Colors.redAccent),
                                              title: const Text('Delete bookmark'),
                                              onTap: () => Navigator.pop(context, true),
                                            ),
                                            ListTile(
                                              leading: Icon(MdiIcons.close, color: tc.textMuted),
                                              title: const Text('Cancel'),
                                              onTap: () => Navigator.pop(context, false),
                                            ),
                                          ],
                                        ),
                                      ),
                                    ),
                                  );
                                  if (confirmed == true) {
                                    _deleteBookmark(b);
                                  }
                                },
                                child: Container(
                                  margin: const EdgeInsets.only(bottom: 10),
                                  padding: const EdgeInsets.all(QuranLayout.cardPadding),
                                  decoration: BoxDecoration(
                                    color: tc.card,
                                    borderRadius: BorderRadius.circular(QuranLayout.cardRadius),
                                    border: Border.all(color: tc.cardBorder),
                                  ),
                                  child: Row(
                                    crossAxisAlignment: CrossAxisAlignment.start,
                                    children: [
                                      Icon(MdiIcons.bookmark, color: tc.accent),
                                      const SizedBox(width: 12),
                                      Expanded(
                                        child: Column(
                                          crossAxisAlignment: CrossAxisAlignment.start,
                                          children: [
                                            Text(
                                              '${surah?.englishName ?? 'Surah ${b.surahNumber}'} · Ayah ${b.ayahNumber}',
                                              style: AppTypography.body(tc).copyWith(fontWeight: FontWeight.w600),
                                            ),
                                            if (preview.isNotEmpty) ...[
                                              const SizedBox(height: 4),
                                              Text(
                                                preview,
                                                style: AppTypography.caption(tc),
                                                textDirection: TextDirection.rtl,
                                                maxLines: 1,
                                                overflow: TextOverflow.ellipsis,
                                              ),
                                            ],
                                          ],
                                        ),
                                      ),
                                      Icon(MdiIcons.chevronRight, color: tc.textMuted),
                                    ],
                                  ),
                                ),
                              );
                            },
                          ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

File: prayer_app_expo/src/screens/QuranBookmarksScreen.js
import React, { useEffect, useState } from 'react';
import { View, Text, StyleSheet, TouchableOpacity, FlatList, ActivityIndicator, Alert } from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { useTheme } from '../providers/ThemeProvider';
import { QuranLayout, getTypography, interFont } from '../theme/theme';
import { loadBookmarks, removeBookmark } from '../services/quranStorageService';
import { fetchSurahList, loadCachedSurahList, loadCachedArabic, fetchSurahArabic } from '../services/quranApi';

export default function QuranBookmarksScreen({ onBack, onOpenReader }) {
    const { theme: tc } = useTheme();
    const typo = getTypography(tc);

    const [bookmarks, setBookmarks] = useState([]);
    const [surahMap, setSurahMap] = useState({});
    const [previews, setPreviews] = useState({});
    const [loading, setLoading] = useState(true);

    useEffect(() => {
        load();
    }, []);

    async function load() {
        setLoading(true);
        const b = await loadBookmarks();

        let surahs = await loadCachedSurahList();
        if (!surahs?.length) {
            try {
                surahs = await fetchSurahList();
            } catch (_) {
                surahs = [];
            }
        }

        const map = {};
        for (const s of surahs) map[s.number] = s;

        setBookmarks(b);
        setSurahMap(map);
        setLoading(false);

        loadPreviews(b);
    }

    async function loadPreviews(bookmarksList) {
        const next = {};
        const cachedSurahAyahs = {};

        for (const b of bookmarksList) {
            if (!cachedSurahAyahs[b.surahNumber]) {
                let ayahs = await loadCachedArabic(b.surahNumber);
                if (!ayahs?.length) {
                    try {
                        ayahs = await fetchSurahArabic(b.surahNumber);
                    } catch (_) {
                        ayahs = [];
                    }
                }
                cachedSurahAyahs[b.surahNumber] = ayahs;
            }

            const ayahs = cachedSurahAyahs[b.surahNumber] || [];
            const idx = b.ayahNumber - 1;
            if (idx >= 0 && idx < ayahs.length) {
                const clean = (ayahs[idx].textAr || '').replace(/\n/g, ' ').trim();
                next[`${b.surahNumber}_${b.ayahNumber}`] = clean.length > 30 ? `${clean.slice(0, 30)}…` : clean;
            } else {
                next[`${b.surahNumber}_${b.ayahNumber}`] = '';
            }
        }

        setPreviews(next);
    }

    function onLongPressDelete(item) {
        Alert.alert(
            'Delete bookmark',
            `Remove ${surahMap[item.surahNumber]?.englishName || `Surah ${item.surahNumber}`} · Ayah ${item.ayahNumber}?`,
            [
                { text: 'Cancel', style: 'cancel' },
                {
                    text: 'Delete',
                    style: 'destructive',
                    onPress: async () => {
                        await removeBookmark(item);
                        load();
                    },
                },
            ]
        );
    }

    return (
        <View style={{ flex: 1 }}>
            <View style={styles.topBar}>
                <TouchableOpacity onPress={onBack} hitSlop={10}>
                    <MaterialCommunityIcons name="arrow-left" size={24} color={tc.textPrimary} />
                </TouchableOpacity>
                <View style={{ width: 12 }} />
                <Text style={typo.titleMedium}>Bookmarks</Text>
            </View>

            {loading ? (
                <View style={styles.center}><ActivityIndicator size="large" color={tc.accent} /></View>
            ) : bookmarks.length === 0 ? (
                <View style={styles.center}><Text style={typo.caption}>No bookmarks yet</Text></View>
            ) : (
                <FlatList
                    data={bookmarks}
                    keyExtractor={(item, idx) => `${item.surahNumber}_${item.ayahNumber}_${idx}`}
                    contentContainerStyle={{ paddingHorizontal: 20, paddingTop: 10, paddingBottom: 20 }}
                    renderItem={({ item }) => {
                        const surah = surahMap[item.surahNumber];
                        const preview = previews[`${item.surahNumber}_${item.ayahNumber}`] || '';
                        return (
                            <TouchableOpacity
                                style={[styles.row, { backgroundColor: tc.card, borderColor: tc.cardBorder }]}
                                activeOpacity={0.85}
                                onPress={() => onOpenReader(surah || { number: item.surahNumber, englishName: `Surah ${item.surahNumber}`, nameAr: '' }, item.ayahNumber, 'bookmarks')}
                                onLongPress={() => onLongPressDelete(item)}
                            >
                                <MaterialCommunityIcons name="bookmark" size={20} color={tc.accent} style={{ marginTop: 2 }} />
                                <View style={{ width: 12 }} />
                                <View style={{ flex: 1 }}>
                                    <Text style={[typo.body, { fontFamily: interFont('600') }]}>
                                        {(surah?.englishName || `Surah ${item.surahNumber}`)} · Ayah {item.ayahNumber}
                                    </Text>
                                    {preview ? (
                                        <Text style={[typo.caption, { marginTop: 4 }]} numberOfLines={1}>{preview}</Text>
                                    ) : null}
                                </View>
                                <MaterialCommunityIcons name="chevron-right" size={20} color={tc.textMuted} />
                            </TouchableOpacity>
                        );
                    }}
                />
            )}
        </View>
    );
}

const styles = StyleSheet.create({
    topBar: {
        flexDirection: 'row',
        alignItems: 'center',
        paddingHorizontal: 12,
        paddingVertical: 8,
    },
    center: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
    },
    row: {
        padding: QuranLayout.cardPadding,
        borderRadius: QuranLayout.cardRadius,
        borderWidth: 1,
        marginBottom: 10,
        flexDirection: 'row',
        alignItems: 'flex-start',
    },
});
